### ISSUE TITLE
ASCII box drawing character(s) not full height

### ISSUE THREAD

**Issue by alerque** [chrissimpkins/Hack#104]
Using the OTF font on Linux I don't have a problem, but on OS-X using the OTF fonts screws up my vim and tmux splits. These are drawn using the `U+2502` vertical line (│) character. This character should be the full height of the line, but using the default line height settings in both Terminal and iTerm it comes up short. It's not unless I compress the line height to ~80% of default that the line connects.

![2.013/OTF/OSX 10.11/iTerm 2](https://cloud.githubusercontent.com/assets/173595/9873099/3b397370-5ba6-11e5-823a-c6779ba4f96b.png)

Again it works fine on Linux using default line heights.

![2.013/OTF/Linux/Termite](https://cloud.githubusercontent.com/assets/173595/9872964/5d87962e-5ba5-11e5-97b9-e02497717cd5.png)

**Comment by chrissimpkins** [comment]
Thanks Caleb. These box glyphs definitely need some attention (along with others like line glyphs that require alignment in the vertical dimension with each other).  I am working on some other issues with the vertical metrics which may lead to minor vertical spacing changes and want to sort that issue out before I work on these issues but will definitely address this.  Interesting that it isn't a problem on Linux.   What Linux distro is that?

**Comment by alerque** [comment]
I'm running a very up to date ArchLinux box, and that is Termite (a VTE3-ng based terminal).

**Comment by chrissimpkins** [comment]
got it thanks.  The reason for the comment above is that any change to the overall vertical metrics of the fonts will scrap changes to the alignment of specific glyphs that we make here.  I will definitely address this.  Hope to have these vertical metrics issues sorted out over the course of this week.

**Comment by chrissimpkins** [comment]
arch is running FreeType?  any modifications (eg. Infinality)?

**Comment by alerque** [comment]
Yes, freetype 2.6, harfbuzz 1.0.3, and graphite 1.3.2. No no infinality or other mods that I know of.

**Comment by chrissimpkins** [comment]
Please give the ttf builds with brand new vertical metrics a shot.  Available in #111

**Comment by chrissimpkins** [comment]
This is what I am seeing for U+2502.  It is well outside of the bounding box so we may have the vertical spacing just above the level where these connect in the current v2.014 test build (as @weslem confirmed).

<img width="476" alt="vertical-bar" src="https://cloud.githubusercontent.com/assets/4249591/10099826/563fb046-635b-11e5-8079-dccdb9be2c1e.png">

Will work on the glyph to see if we can address this.

**Comment by weslem** [comment]
If it helps, here's the current rendering of the test build captured from my system in `st` with `Hack:size=10:antialias=true`:

![columns](https://cloud.githubusercontent.com/assets/994468/10105238/711fdd62-6362-11e5-9e4f-76b72f3ca2c4.png)

**Comment by chrissimpkins** [comment]
@weslem do you happen to have DejaVu Sans Mono installed to check where it is with their metrics as well?   I didn't (intentionally) change this glyph and the vertical metrics should be close to their set as of the v2.014 changes.

**Comment by chrissimpkins** [comment]
@weslem and thanks!

**Comment by weslem** [comment]
Sorry, I somehow missed your request. Here's some captures:

```shell
xterm -fa 'Hack:size=12' -bg black -fg white -e \
  'for x in {0..10}; do echo xX│Xx; done; read'`
```
![hack](https://cloud.githubusercontent.com/assets/994468/10227323/8663401a-6821-11e5-820b-89a591ecb0ce.png)

```shell
xterm -fa 'DejaVu Sans Mono:size=12' -bg black -fg white -e \
  'for x in {0..10}; do echo xX│Xx; done; read'
```
![dejavu](https://cloud.githubusercontent.com/assets/994468/10227310/70202340-6821-11e5-8b9b-41cd76b6f44f.png)

**Comment by chrissimpkins** [comment]
Interesting, not sure why they differ. I will take a look at the DejaVu glyph.  Thanks for posting this.

**Comment by weslem** [comment]
I took a look at this on some of my other machines. All of them have connected box-line characters, as well as a shared "dotfile" configuration repo between them, making the differences easier to find.

The only configuration difference I found is that the quirky machine has a 101 DPI monitor. X had trouble identifying the DPI, and kept setting it to 96 regardless of any change I made to Xorg.conf, so I had to manually set it in my fontconfig. It ends up with a weird configuration where X still thinks it's 96 dpi:

```shell
> xdpyinfo | grep dots
  resolution:    96x96 dots per inch
```

But fontconfig thinks it's 101 DPI:

```xml
 <match target="pattern">
   <edit name="dpi" mode="assign">
     <double>101</double>
   </edit>
 </match>
```

Given the complexity of Linux, I have no confidence that this is the only difference, but it seemed suspicious to me.

**Comment by chrissimpkins** [comment]
Oh that's interesting.  Are they showing as aligned in the Hack set on the other machines?

**Comment by weslem** [comment]
I'm not sure what you mean when you ask, "showing as aligned." The vertical line box chars definitely connect, but I haven't done any deeper investigation.

I *have* manually set the DPI to 96 in `fonts.conf`, with no visible improvement in vertical alignment. Removing my manual DPI setting completely produces a minuscule text size (surprising me, given that xdpyinfo says X has the 96 DPI setting), but it still has an empty pixel between the lines of characters.

To round out my diagnostics, I've looked into library versions on the machines. The bad machine is on Ubuntu, my two good machines are Ubuntu and an Archlinux box. Both ubuntu machines have the 2.11.0-ubuntu4.1 amd64 version of fontconfig.

I'll follow-up with a post from that machine with some more screen grabs that show slight differences in rendering. It looks like the "good" Ubuntu machine may still have some slight offset that is hard to see due to antialiasing.

**Comment by weslem** [comment]
On the "good" ubuntu machine, xterm seems to render things perfectly:

```shell
xterm -fa 'Hack:size=12' -bg black -fg white -e \
  'for x in {0..10}; do echo xX│Xx; done; read'
```
![xterm](https://cloud.githubusercontent.com/assets/994468/10493403/c93dcd62-7265-11e5-99f4-377961a3e5f7.png)


But the `st` shell from which I normally operate, with an alpha-blended background, shows slight gaps filled in with antialiasing even for the same `Hack:size=12` font string:
![st](https://cloud.githubusercontent.com/assets/994468/10493274/196f8952-7265-11e5-8a07-4df54e5ad736.png)

**Comment by chrissimpkins** [comment]
We are going to include these changes in v2.020.  Currently working on the block element glyphs for v2.019.

**Comment by 15348805850hanwannian** [comment]
韩万年

**Comment by chrissimpkins** [comment]
v2.020 DEV012616 build files are available for testing:

[Hack-Regular-DEV.ttf](https://github.com/chrissimpkins/Hack/raw/development/build/test_builds/Hack-Regular-DEV.ttf)

[Hack-Italic-DEV.ttf](https://github.com/chrissimpkins/Hack/raw/development/build/test_builds/Hack-Italic-DEV.ttf)

[Hack-Bold-DEV.ttf](https://github.com/chrissimpkins/Hack/raw/development/build/test_builds/Hack-Bold-DEV.ttf)

[Hack-BoldItalic-DEV.ttf](https://github.com/chrissimpkins/Hack/raw/development/build/test_builds/Hack-BoldItalic-DEV.ttf)

**Comment by weslem** [comment]
I still produce strange renderings on my high-DPI monitor. I made a number of renderings with both 2.019 and the DEV012616 version using this script:

```shell
for size in {8..14};
  do xterm -geometry 5x11 -fa "Hack:size=$size" -bg black -fg white -e \
    "echo ${size}pt; for x in {0..8}; 
      do echo xX│Xx;
     done;
    import /tmp/2_20_DEV-size${size}.png";
done
```

Point size 12 looks good, but does show slight differences around the vertical bar meeting points:
![12_summary](https://cloud.githubusercontent.com/assets/994468/12588340/1c01d82e-c40d-11e5-8d55-e6e6117781ac.png)

Point size 11 is visibly different, with a clear gap in both versions. The v2.020 DEV012616 version seems to have a smaller gap, however:

![11_summary](https://cloud.githubusercontent.com/assets/994468/12588364/30a48344-c40d-11e5-8052-ce73f8455729.png)

I'm the only one that continues to comment on this spacing issue. Unless others observe it as well, it seems like it shouldn't be a priority.

**Comment by chrissimpkins** [comment]
Thanks Matthew (@weslem).  I really appreciate your feedback.  
 
> Unless others observe it as well, it seems like it shouldn't be a priority.

We have several open issue reports associated with vertical spacing in the box drawing glyph set.  It was work that needed to be done after I adjusted the line spacing.  It looks like we are getting close.  At this point I think that we are entering the domain of hinting issues and I am not prepared to manually modify the hinting across all of these glyphs right now.

Thanks again.

**Comment by chrissimpkins** [comment]
It's a pretty big set of glyphs and here is where we are overall:

<img width="233" alt="block-drawing" src="https://cloud.githubusercontent.com/assets/4249591/12587606/1b98faa0-c423-11e5-8cd3-dae661b69f42.png">

**Comment by chrissimpkins** [comment]
These fixes (our best attempt at them) were included in the v2.020 release.  Now available.  Thanks to all for your help with this issue.
